### ISSUE TITLE
test_analyzer_resnet50 fails on E5-2660 v4 machine

### ISSUE THREAD

**Issue by luotao1** [PaddlePaddle/Paddle#18005]
### log
total: http://ci.paddlepaddle.org/viewType.html?buildTypeId=Paddle_PrCiNightCuda10cudnn7develop&branch_Paddle=%3Cdefault%3E&tab=buildTypeStatusDiv
```
[01:17:01] 71/295 Test #155: test_analyzer_resnet50 .......................***Exception: Illegal  1.40 sec
[01:17:01][==========] Running 8 tests from 1 test case.
[01:17:01][----------] Global test environment set-up.
[01:17:01][----------] 8 tests from Analyzer_resnet50
[01:17:01][ RUN      ] Analyzer_resnet50.profile
[01:17:01]WARNING: Logging before InitGoogleLogging() is written to STDERR
[01:17:01]I0611 01:17:00.184201 24524 tester_helper.h:233] feed target 0: {-1, 3, 318, 318}
[01:17:01]I0611 01:17:00.187963 24524 tester_helper.h:87] AnalysisConfig {
```
log1: http://ci.paddlepaddle.org/viewLog.html?tab=buildLog&buildTypeId=Paddle_PrCiNightCuda10cudnn7develop&buildId=112341&_focus=26941
log2: http://ci.paddlepaddle.org/viewLog.html?tab=buildLog&buildTypeId=Paddle_PrCiNightCuda10cudnn7develop&buildId=111328&_focus=26255
...
### machine
![image](https://user-images.githubusercontent.com/6836917/59269133-e22fd980-8c80-11e9-9b46-76543ff4d70d.png)

**Comment by luotao1** [comment]
Does it related with #15226, i.e 
https://github.com/PaddlePaddle/Paddle/blob/0c39b97b4ee0f732a6a9a349511b3ac3cdc1633c/paddle/fluid/inference/tests/api/CMakeLists.txt#L37-L39

**Comment by luotao1** [comment]
I can reproduce on my local avx2 machine.
```
diff --git a/paddle/fluid/inference/tests/api/CMakeLists.txt b/paddle/fluid/inference/tests/api/CMakeLists.txt
index 548ef38..d99d2f9 100644
--- a/paddle/fluid/inference/tests/api/CMakeLists.txt
+++ b/paddle/fluid/inference/tests/api/CMakeLists.txt
@@ -149,7 +149,7 @@ inference_analysis_api_test_with_fake_data(test_analyzer_googlenet

 # resnet50
 inference_analysis_api_test_with_fake_data(test_analyzer_resnet50
-  "${INFERENCE_DEMO_INSTALL_DIR}/resnet50" analyzer_resnet50_tester.cc "resnet50_model.tar.gz" true)
+  "${INFERENCE_DEMO_INSTALL_DIR}/resnet50" analyzer_resnet50_tester.cc "resnet50_model.tar.gz" false)
```
results:
```
130: ---  detected 1 subgraphs
130: --- Running analysis [ir_params_sync_among_devices_pass]
130: --- Running analysis [adjust_cudnn_workspace_size_pass]
130: --- Running analysis [ir_graph_to_program_pass]
130: I0611 14:06:44.485680 14470 analysis_predictor.cc:451] == optimize end ==
130: I0611 14:06:44.486610 14470 tester_helper.h:331] Thread 0, number of threads 1, run 1 times...
130: E0611 14:06:44.487390 14470 analysis_predictor.cc:303] feed names from program do not have name: [] from specified input
130: I0611 14:06:44.762948 14470 helper.h:322] ====== threads: 1, thread id: 0 ======
130: I0611 14:06:44.762974 14470 helper.h:324] ====== batch size: 1, iterations: 1, repetitions: 1 ======
130: I0611 14:06:44.762980 14470 helper.h:326] ====== batch latency: 276.315ms, number of samples: 1, sample latency: 276.315ms, fps: 3.61905, data type: float ======
130: F0611 14:06:44.769438 14470 tester_helper.h:119] Check failed: std::abs(pdata_ref[j] - pdata[j]) <= FLAGS_accuracy (0.00100708 vs. 0.001)
130: *** Check failure stack trace: ***
1/1 Test #130: test_analyzer_resnet50 ...........***Exception: Other  4.49 sec
```

**Comment by wojtuss** [comment]
@luotao1 
Yes, it is related to https://github.com/PaddlePaddle/Paddle/pull/15226.
The machine you are using does not support AVX512 instructions, and it supports AVX2 only. Thus, forcing MKL to use AVX512 instructions (by setting the `mkl_debug` argument to `true` in the `inference_analysis_api_test_with_fake_data` for the `test_analyzer_resnet50` test, see `paddle/fluid/inference/tests/api/CMakeLists.txt:152`) must fail.
If the `mkl_debug` argument were set to `false`, the test would also fail by exceeding the output difference threshold (0.001) (due to using AVX2 instructions instead of AVX512), exactly as it was already discussed in comments to the PR https://github.com/PaddlePaddle/Paddle/pull/15226.

IMO, if a machine supporting only AVX2 had to be used for CI, a solution to this issue could be to increase the output difference threshold (and drop the workaround with the `mkl_debug=true` argument as not needed anymore) or to use different output pattern for AVX2-based calculations.

@Sand3r- , can you confirm that?

**Comment by Sand3r-** [comment]
Totally, yes. We don't have much influence of how MKL's avx2 gemm is performing its calculations.
An alternative that I see that is harder to attain, would be to find such reference output that would make the test pass for both avx2 and avx512. But in my opinion increasing the output difference treshold shall suffice.

**Comment by wojtuss** [comment]
@luotao1 ,
For now, we will:
1. disable the `fc_mkldnn_pass` for ResNet50 test,
2. remove forcing AVX512 instructions by `mkl_debug=true`.

Then we will investigate for more general solution to this issue.

**Comment by luotao1** [comment]
I find this test core dump on AVX machine.
```
$ctest -R test_analyzer_resnet -V
UpdateCTestConfiguration  from :/home/luotao/Paddle/cpu_build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/luotao/Paddle/cpu_build/DartConfiguration.tcl
Test project /home/luotao/Paddle/cpu_build
Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 130
    Start 130: test_analyzer_resnet50

130: Test command: /home/luotao/Paddle/cpu_build/paddle/fluid/inference/tests/api/test_analyzer_resnet50 "--inference_model_dir=/home/luotao/Paddle/cpu_build/third_party/inference_demo/word2vec/word2vec.inference.model" "--infer_model=/home/luotao/Paddle/cpu_build/third_party/inference_demo/resnet50/model"
130: Environment variables:
130:  FLAGS_cudnn_deterministic=true
130:  MKL_DEBUG_CPU_TYPE=7
130: Test timeout computed to be: 600
130: [==========] Running 8 tests from 1 test case.
130: [----------] Global test environment set-up.
130: [----------] 8 tests from Analyzer_resnet50
130: [ RUN      ] Analyzer_resnet50.profile
130: WARNING: Logging before InitGoogleLogging() is written to STDERR
130: I0613 12:51:32.969542 20477 tester_helper.h:233] feed target 0: {-1, 3, 318, 318}
130: I0613 12:51:32.974023 20477 tester_helper.h:87] AnalysisConfig {
130:   NativeConfig {
130:     PaddlePredictor::Config {
130:       model_dir:
130:     }
130:     use_gpu: 0
130:     device: 0
130:     fraction_of_gpu_memory: 0
130:     specify_input_name: 1
130:   }
130:   prog_file: /home/luotao/Paddle/cpu_build/third_party/inference_demo/resnet50/model/model
130:   param_file: /home/luotao/Paddle/cpu_build/third_party/inference_demo/resnet50/model/params
130:   enable_ir_optim: 1
130:   cpu_num_threads: 1
130:   use_feed_fetch_ops: 1
130:   use_tensorrt: 0
130:   use_mkldnn: 0
130:   use_ngraph: 0
130: }
130: --- Running analysis [ir_graph_build_pass]
130: --- Running analysis [ir_analysis_pass]
130: --- Running IR pass [infer_clean_graph_pass]
130: --- Running IR pass [attention_lstm_fuse_pass]
130: --- Running IR pass [seqconv_eltadd_relu_fuse_pass]
130: --- Running IR pass [fc_lstm_fuse_pass]
130: --- Running IR pass [mul_lstm_fuse_pass]
130: --- Running IR pass [fc_gru_fuse_pass]
130: --- Running IR pass [mul_gru_fuse_pass]
130: --- Running IR pass [seq_concat_fc_fuse_pass]
130: --- Running IR pass [fc_fuse_pass]
130: ---  detected 1 subgraphs
130: --- Running IR pass [repeated_fc_relu_fuse_pass]
130: --- Running IR pass [squared_mat_sub_fuse_pass]
130: --- Running IR pass [conv_bn_fuse_pass]
130: --- Running IR pass [conv_eltwiseadd_bn_fuse_pass]
130: ---  detected 53 subgraphs
130: --- Running IR pass [is_test_pass]
130: --- Running IR pass [runtime_context_cache_pass]
130: --- Running analysis [ir_params_sync_among_devices_pass]
130: --- Running analysis [adjust_cudnn_workspace_size_pass]
130: --- Running analysis [ir_graph_to_program_pass]
130: I0613 12:51:33.735329 20477 analysis_predictor.cc:451] == optimize end ==
130: I0613 12:51:33.737641 20477 tester_helper.h:331] Thread 0, number of threads 1, run 1 times...
130: E0613 12:51:33.739043 20477 analysis_predictor.cc:303] feed names from program do not have name: [] from specified input
1/1 Test #130: test_analyzer_resnet50 ...........***Exception: Illegal  1.88 sec
```
- machine:
```
processor	: 10
vendor_id	: GenuineIntel
cpu family	: 6
model		: 62
model name	: Intel(R) Xeon(R) CPU E5-2620 v2 @ 2.10GHz
stepping	: 4
microcode	: 0x1
cpu MHz		: 2099.998
cache size	: 4096 KB
physical id	: 10
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 10
initial apicid	: 10
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl eagerfpu pni pclmulqdq ssse3 cx16 pcid sse4_1 sse4_2 x2apic popcnt aes xsave avx f16c rdrand hypervisor lahf_lm arat fsgsbase tsc_adjust smep erms xsaveopt
bogomips	: 4199.99
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:
```

**Comment by ghost** [comment]
@luotao1 Why set environment variable "MKL_DEBUG_CPU_TYPE=7" in above log?

**Comment by luotao1** [comment]
@jianhang-liu Thanks for your reminder! After `make clean & make -j` again, this unit-test runs successfully on AVX machine.

**Comment by bingyanghuang** [comment]
@luotao1 Looks like your problem has been solved. If so , could you help close this PR?

**Comment by luotao1** [comment]
This unit-test is disabled not fixed now, do you have another issue to track it? @bingyanghuang

**Comment by bingyanghuang** [comment]
@luotao1 I think no. @wojtuss @Sand3r- Do you have any new comments or updates on this issue?

**Comment by wojtuss** [comment]
@luotao1 , @bingyanghuang 
We will try to reproduce the issue on our side and look for a solution for it by the end of WW31

**Comment by bingyanghuang** [comment]
> @luotao1 , @bingyanghuang
> We will try to reproduce the issue on our side and look for a solution for it by the end of WW31

This week is WW27, so WW31 is about the beginning of the August.

**Comment by Sand3r-** [comment]
Thanks to @grygielski  We have managed to reproduce the problem on our AVX2 machines on develop branch:
```
develop: F0910 08:07:30.742866    20 tester_helper.h:119] Check failed: std::abs(pdata_ref[j] - pdata[j]) <= FLAGS_accuracy (0.00100708 vs. 0.001)
```

as well as on MKL-DNN 1.0 to see if the problem persists (unfortunately it does):
```
F0910 08:07:45.358696    21 tester_helper.h:119] Check failed: std::abs(pdata_ref[j] - pdata[j]) <= FLAGS_accuracy (0.0010376 vs. 0.001)
```

We believe that it is caused by using AVX2 version of MKL's gemm. We will attempt to investigate the problem furtherly.

**Comment by grygielski** [comment]
It all comes to checking condition on `CHECK_LE(std::abs(pdata_ref[j] - pdata[j]), FLAGS_accuracy);` in compare_mkldnn part of analyzer_resnet50 test. It checks if every value of output tensor doesn't differ more than 0.001 from reference one.  I guess it supposed to be 0.1% difference threshold on model prediction accuracy but model in this test doesn't have softmax activation function at the end. Output tensor has values from -1183.647994 to +1050.651855 so looking for 0.001 difference between them is really overprecise. Model path used in test: `"build_folder"/third_party/inference_demo/resnet50/model`.  It's really random that this test passes now for AVX512 but does not for AVX2. 

![image](https://user-images.githubusercontent.com/38704900/64611769-7c173780-d3d2-11e9-9d4a-60e2131a998a.png)

As you can see this test would even fail if we compare reference value from AVX2 and AVX512 and this difference is still 3x bigger than MKLDNN difference. All these values were calculated with flag disable_mkldnn_fc=false. Disabling FC mkldnn op randomly made the test pass but the problem is in lacking softmax operator since relative error is about 5e-7 in this case. Please refer to my conclusions and ask questions if there are any. @luotao1 @bingyanghuang

**Comment by bingyanghuang** [comment]
@grygielski Based on your analysis, if I understand correctly , this error is a random diff which possibly happens both in AVX2 and AVX512? You mentioned that the problem is lacking softmax operator, what do you mean ?

**Comment by grygielski** [comment]
This test fails only on AVX2 but the reason is pretty random. Test calculates difference between ref value (without passes and other modifications to graph i guess) and tested graph setup. It happens that MKLDNN implementation of whole graph differs very little between AVX2 and AVX512 and ref value between machines changes 3x more. This is what causes test fails.

But the main problem here is that Resnet architecture ends with FC layer that should give vector of predictions as an output. This vector should consists of values between 0 and 1 indicating how sure neural network is that input image is of certain class (e.g. 0.1 = 10%) and all output values summing to 1 (which is 100% total). This is what softmax activation does. With such values we can assert that value of prediction should not differ more than 0.001 (current test threshold) what means that for every class prediction won't change more than 0.1 percentage point from reference one. Right now in the model it is just output from FC layer without any activation function so output vector values are between -1183.647994 and +1050.651855. With such large numbers, looking for 0.001 difference should not take place since it is really tiny relative difference and it won't affect prediction at all. @bingyanghuang

**Comment by Sand3r-** [comment]
Has the issue been resolved and can it be closed? @luotao1

**Comment by bingyanghuang** [comment]
We have proposed the solution to this issue, I think we can close this issue from Intel side.
